### PR TITLE
Preserve "blocks" set when updating from cache

### LINKED ISSUE #845: A cached file once marked as sparse remains so after reading everything
If a file in a caching system (for the HTTP backend, at least) is opened, closed without reading everything, opened again, and read to the end, subsequent opens of the same file will continue to treat it as sparsely cached and will refetch one or more blocks despite the fact that everything should have been fetched during the second "open" session.  ~~Unlike #799, this happens for files both smaller and larger than 5 MiB.~~ (EDIT: It turns out the URL I was using puts a limit of 102400 on the returned file size, so I have not actually tested the code with a > 5 MiB file.)

MVCE:

```python
from hashlib import sha256
import logging
import tempfile
import fsspec
from fsspec.implementations.cached import CachingFileSystem

logging.basicConfig(
    format="%(asctime)s [%(levelname)-8s] %(name)s: %(message)s",
    datefmt="%Y-%m-%dT%H:%M:%S%z",
    level=logging.DEBUG,
)

FILESIZE = 1_000_000
BLOCK_SIZE = 65536

with tempfile.TemporaryDirectory() as tmpdir:
    fs = CachingFileSystem(fs=fsspec.filesystem("http"), cache_storage=tmpdir)

    with fs.open(f"https://httpbin.org/bytes/{FILESIZE}") as f:
        pass

    with fs.open(f"https://httpbin.org/bytes/{FILESIZE}") as f:
        m = sha256()
        for blob in iter(lambda: f.read(BLOCK_SIZE), b""):
            m.update(blob)
        print("***", m.hexdigest())

    with fs.open(f"https://httpbin.org/bytes/{FILESIZE}") as f:
        m = sha256()
        for blob in iter(lambda: f.read(BLOCK_SIZE), b""):
            m.update(blob)
        print("***", m.hexdigest())
```

Running this script produces:

```
2021-12-07T16:16:22-0500 [DEBUG   ] asyncio: Using selector: KqueueSelector
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec: Creating local sparse file for https://httpbin.org/bytes/1000000
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec.http: Retrieve file size for https://httpbin.org/bytes/1000000
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec: going to save
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec: saved
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec: Opening partially cached copy of https://httpbin.org/bytes/1000000
2021-12-07T16:16:22-0500 [DEBUG   ] fsspec.http: Retrieve file size for https://httpbin.org/bytes/1000000
2021-12-07T16:16:23-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 0 - 65536
2021-12-07T16:16:23-0500 [DEBUG   ] fsspec: MMap cache fetching 0-65536
2021-12-07T16:16:23-0500 [DEBUG   ] fsspec: MMap get block #0 (0-102400
2021-12-07T16:16:23-0500 [DEBUG   ] fsspec.http: Fetch range for <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000>: 0-102400
2021-12-07T16:16:23-0500 [DEBUG   ] fsspec.http: https://httpbin.org/bytes/1000000 : bytes=0-102399
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 65536 - 102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: MMap cache fetching 65536-102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 102400 - 102400
*** 53f743e75fba805e31f176b0ddc15dc624d23d50e29b850ea3ce851136b803e6
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: going to save
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: saved
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: Opening partially cached copy of https://httpbin.org/bytes/1000000
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec.http: Retrieve file size for https://httpbin.org/bytes/1000000
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 0 - 65536
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: MMap cache fetching 0-65536
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: MMap get block #0 (0-102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec.http: Fetch range for <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000>: 0-102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec.http: https://httpbin.org/bytes/1000000 : bytes=0-102399
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 65536 - 102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: MMap cache fetching 65536-102400
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 102400 - 102400
*** 5fa5cdb53afa8daa682d526b208f71b01cb618e87146e91d2890a9dd6c9def17
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: going to save
2021-12-07T16:16:24-0500 [DEBUG   ] fsspec: saved
```

Note that, when the URL is opened for the third time, fsspec says it is "partially cached" despite the fact that everything should have been read during the second "open" block.  As a result, the contents of the URL are fetched again, and because the URL is dynamic, a different digest is obtained.

Compare this to the output if the `with fs.open(...) as f: pass` block is commented out:

```
2021-12-07T16:16:35-0500 [DEBUG   ] asyncio: Using selector: KqueueSelector
2021-12-07T16:16:35-0500 [DEBUG   ] fsspec: Creating local sparse file for https://httpbin.org/bytes/1000000
2021-12-07T16:16:35-0500 [DEBUG   ] fsspec.http: Retrieve file size for https://httpbin.org/bytes/1000000
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 0 - 65536
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: MMap cache fetching 0-65536
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: MMap get block #0 (0-102400
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec.http: Fetch range for <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000>: 0-102400
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec.http: https://httpbin.org/bytes/1000000 : bytes=0-102399
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 65536 - 102400
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: MMap cache fetching 65536-102400
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: <File-like object HTTPFileSystem, https://httpbin.org/bytes/1000000> read: 102400 - 102400
*** b62647877d5679edd69048192afd1b57331800fff24894d6ffec97a3013de064
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: going to save
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: saved
2021-12-07T16:16:36-0500 [DEBUG   ] fsspec: Opening local copy of https://httpbin.org/bytes/1000000
*** b62647877d5679edd69048192afd1b57331800fff24894d6ffec97a3013de064
```

Here, fsspec correctly caches everything when it is all read during the first "open" block, and so the digests match.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #845.

### CODE REVIEW

File: fsspec/implementations/tests/test_cached.py [hunk 1]
```diff
@@ -578,10 +578,12 @@ def test_cached_open_close_read(ftp_writable):
...
         target_options={"host": host, "port": port, "username": user, "password": pw},
     )
-    with fs.open("/out_block") as f:
+    with fs.open("/out_block", block_size=1024) as f:
         pass
-    with fs.open("/out_block") as f:
+    with fs.open("/out_block", block_size=1024) as f:
         assert f.read(1) == b"t"
+    # Regression test for <https://github.com/fsspec/filesystem_spec/issues/845>
+    assert fs.cached_files[-1]["/out_block"]["blocks"] == {0}
```
Comment (martindurant) [inline fsspec/implementations/tests/test_cached.py:586]: Does this test fail without the code change in cached.py?

Comment (jwodder) [reply fsspec/implementations/tests/test_cached.py:586]: Yes, it does fail.

Comment (martindurant) [reply fsspec/implementations/tests/test_cached.py:586]: Perfect, thanks.

**Comment by martindurant** [comment]
Thank you, @jwodder

**Pull request merged**